core/common/scale.cpp

```diff
@@ -0,0 +1,493 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "scale.hpp"
+
+#include <assert.h>
+#include <endian.h>
+#include <limits>
+#include <tuple>
+#include <type_traits>
+#include <algorithm>
+
+namespace kagome::common {
+
+  namespace scale {
+    namespace util {
+      /**
+       * @brief Union class that helps representing integer value as an array
+       * @tparam T Integer type
+       */
+      template <class T>
+      union AsArray {
+        constexpr explicit AsArray(const T &t) : value_{t} {}
+
+        ///
+        constexpr uint8_t *begin() {
+          return static_cast<uint8_t *>(array_);
+        }
+
+        constexpr uint8_t *end() {
+          return begin() + size();
+        }
+
+        constexpr static const size_t size() {
+          return sizeof(T);
+        };
+
+        constexpr AsArray reversed() const {
+          AsArray r{value_};
+          std::reverse(r.begin(), r.end());
+          return r;
+        }
+
+        inline auto value() const {
+          return value_;
+        }
+
+        constexpr auto toLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+        constexpr auto fromLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+       private:
+        constexpr inline auto reverseIfHostNotLittleEndian() const {
+          if (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__) {
+            return value();
+          } else {
+            return reversed().value();
```

Comment:
afaik endianness does not change for *most* platforms during runtime, so this can be something like 
```
#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
return value();
#else 
return reversed().value();
#endif

Reply:
I can make runtime check, but it isn't productive check it every time. Maybe singleton-like thing can save the day, but using singletons are pretty ugly.

Reply:
This conversation seems to be resolved. Correct compile time check is used.